File: app/controllers/iop/transfer/impl/DataTransferController.java
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package controllers.iop.transfer.impl;

import be.objectify.deadbolt.java.actions.Group;
import be.objectify.deadbolt.java.actions.Restrict;
import be.objectify.deadbolt.java.actions.SubjectNotPresent;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import controllers.base.BaseController;
import io.ebean.DB;
import io.ebean.text.PathProperties;
import java.io.File;
import java.io.IOException;
import java.net.MalformedURLException;
import java.net.URI;
import java.net.URL;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;
import javax.inject.Inject;
import miscellaneous.config.ConfigReader;
import miscellaneous.file.FileHandler;
import miscellaneous.json.JsonDeserializer;
import models.attachment.Attachment;
import models.questions.Question;
import models.questions.Tag;
import models.user.User;
import org.apache.pekko.stream.IOResult;
import org.apache.pekko.stream.javadsl.FileIO;
import org.apache.pekko.stream.javadsl.Source;
import org.apache.pekko.util.ByteString;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import play.http.HttpErrorHandler;
import play.libs.Files;
import play.libs.Json;
import play.libs.ws.WSClient;
import play.libs.ws.WSRequest;
import play.mvc.BodyParser;
import play.mvc.Http;
import play.mvc.Result;
import play.mvc.Results;
import security.Authenticated;
import validation.core.Attrs;

public class DataTransferController extends BaseController {

    private final Logger logger = LoggerFactory.getLogger(DataTransferController.class);

    static class DataTransferBodyParser extends BodyParser.Json {

        private static final int SEVENTY_MB = 70000 * 1024;

        @Inject
        DataTransferBodyParser(HttpErrorHandler errorHandler) {
            super(SEVENTY_MB, errorHandler);
        }
    }

    enum DataType {
        QUESTION,
    }

    private final WSClient wsClient;
    private final ConfigReader configReader;
    private final FileHandler fileHandler;

    @Inject
    DataTransferController(WSClient wsClient, ConfigReader configReader, FileHandler fileHandler) {
        this.wsClient = wsClient;
        this.configReader = configReader;
        this.fileHandler = fileHandler;
    }

    @SubjectNotPresent
    public Result importQuestionAttachment(Long id, Http.Request request) {
        Question question = DB.find(Question.class, id);
        if (question == null) {
            return notFound();
        }
        Http.MultipartFormData<Files.TemporaryFile> body = request.body().asMultipartFormData();
        Http.MultipartFormData.FilePart<Files.TemporaryFile> filePart = body.getFile("file");
        if (filePart == null) {
            throw new IllegalArgumentException("file not found");
        }
        if (filePart.getFileSize() > configReader.getMaxFileSize()) {
            throw new IllegalArgumentException("i18n_file_too_large");
        }
        String newFilePath;
        try {
            newFilePath = copyFile(filePart.getRef(), "question", Long.toString(id));
        } catch (IOException e) {
            return internalServerError("i18n_error_creating_attachment");
        }
        // Remove the existing one if found
        fileHandler.removePrevious(question);
        Attachment attachment = fileHandler.createNew(filePart, newFilePath);
        question.setAttachment(attachment);
        question.save();
        return created();
    }

    private String copyFile(Files.TemporaryFile srcFile, String... pathParams) throws IOException {
        String newFilePath = fileHandler.createFilePath(pathParams);
        fileHandler.copyFile(srcFile, new File(newFilePath));
        return newFilePath;
    }

    @SubjectNotPresent
    @BodyParser.Of(DataTransferBodyParser.class)
    public Result importData(Http.Request request) {
        JsonNode body = request.body().asJson();
        if (body.get("type").asText().equals(DataType.QUESTION.toString())) {
            return importQuestions(body);
        }
        return notAcceptable();
    }

    @Authenticated
    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    public CompletionStage<Result> exportData(Http.Request request) throws IOException {
        JsonNode body = request.body().asJson();
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        if (body.get("type").asText().equals(DataType.QUESTION.toString()) && !body.withArray("ids").isEmpty()) {
            String path = "/integration/iop/import";
            Set<Long> ids = StreamSupport.stream(body.get("ids").spliterator(), false)
                .map(JsonNode::asLong)
                .collect(Collectors.toSet());
            PathProperties pp = PathProperties.parse("(*, options(*), tags(name))");
            var query = DB.find(Question.class);
            query.apply(pp);
            Set<Question> questions = query
                .where()
                .idIn(ids)
                .or()
                .eq("questionOwners", user)
                .eq("creator", user)
                .endOr()
                .findSet();

            JsonNode data = ((ObjectNode) body).put("path", path)
                .put("owner", user.getEppn())
                .set(
                    "questions",
                    Json.newArray().addAll(
                        questions
                            .stream()
                            .map(q -> serialize(q, pp))
                            .toList()
                    )
                );

            URL url = parseURL(body.get("orgRef").asText());
            String uploadUrl = parseUploadURL(body.get("orgRef").asText());
            WSRequest wsr = wsClient.url(url.toString());
            return wsr
                .post(data)
                .thenComposeAsync(response -> {
                    JsonNode root = response.asJson();
                    if (response.getStatus() != Http.Status.CREATED) {
                        return wrapAsPromise(internalServerError(root.get("message").asText("Connection refused")));
                    }
                    Map<Long, Long> entries = StreamSupport.stream(root.get("ids").spliterator(), false).collect(
                        Collectors.toMap(id -> id.get("src").asLong(), id -> id.get("dst").asLong())
                    );
                    Map<Long, Attachment> localAttachments = questions
                        .stream()
                        .filter(q -> q.getAttachment() != null && new File(q.getAttachment().getFilePath()).exists())
                        .collect(Collectors.toMap(Question::getId, Question::getAttachment));
                    // Map question copy ids to attachments
                    Map<Long, Attachment> remoteAttachments = localAttachments
                        .entrySet()
                        .stream()
                        .filter(e -> entries.containsKey(e.getKey()))
                        .collect(Collectors.toMap(e -> entries.get(e.getKey()), Map.Entry::getValue));
                    return CompletableFuture.allOf(
                        remoteAttachments
                            .entrySet()
                            .stream()
                            .map(ra -> {
                                String host = uploadUrl.replace("/id/", String.format("/%d/", ra.getKey()));
                                WSRequest req = wsClient.url(host);
                                return CompletableFuture.runAsync(() ->
                                    req
                                        .post(createSource(ra.getValue()))
                                        .exceptionally(e -> {
                                            logger.error("failed in uploading attachment id {}", ra.getKey(), e);
                                            return null;
                                        })
                                );
                            })
                            .toArray(CompletableFuture[]::new)
                    ).thenComposeAsync(_ -> wrapAsPromise(created()));
                });
        }
        return wrapAsPromise(badRequest());
    }

    private Source<Http.MultipartFormData.Part<? extends Source<ByteString, ?>>, ?> createSource(
        Attachment attachment
    ) {
        Source<ByteString, CompletionStage<IOResult>> source = FileIO.fromPath(Path.of(attachment.getFilePath()));
        Http.MultipartFormData.FilePart<Source<ByteString, CompletionStage<IOResult>>> filePart =
            new Http.MultipartFormData.FilePart<>("file", attachment.getFileName(), attachment.getMimeType(), source);
        return Source.from(Set.of(filePart));
    }

    private URL parseURL(String orgRef) throws MalformedURLException {
        String url = String.format("%s/api/organisations/%s/export", configReader.getIopHost(), orgRef);
        return URI.create(url).toURL();
    }

    private String parseUploadURL(String orgRef) throws MalformedURLException {
        String url = String.format("%s/api/organisations/%s/export/id/attachment", configReader.getIopHost(), orgRef);
        return URI.create(url).toURL().toString();
    }

    private boolean isNewTag(Tag tag, List<Tag> existing) {
        return existing.stream().noneMatch(e -> e.getName().equals(tag.getName()));
    }

    private static class QuestionEntry {

        public Long srcId;
        public Long dstId;

        QuestionEntry(Long src, Long dst) {
            this.srcId = src;
            this.dstId = dst;
        }
    }

    private Result importQuestions(JsonNode node) {
        String eppn = node.get("owner").asText();
        Optional<User> ou = DB.find(User.class).where().eq("eppn", eppn).findOneOrEmpty();
        if (ou.isEmpty()) {
            return badRequest("User not recognized");
        }
        User user = ou.get();
        ArrayNode questionNode = node.withArray("questions");
        List<QuestionEntry> entries = StreamSupport.stream(questionNode.spliterator(), false)
            .map(n -> {
                Question question = JsonDeserializer.deserialize(Question.class, n);
                Question copy = question.copy();
                copy.setParent(null);
                copy.setCreatorWithDate(user);
                copy.setModifierWithDate(user);
                copy.save();
                List<Tag> userTags = DB.find(Tag.class).where().eq("creator", user).findList();
                List<Tag> newTags = question
                    .getTags()
                    .stream()
                    .filter(t -> isNewTag(t, userTags))
                    .toList();
                newTags.forEach(t -> t.setId(null));
                List<Tag> existingTags = userTags
                    .stream()
                    .filter(t -> !isNewTag(t, question.getTags()))
                    .toList();
                DB.saveAll(newTags);
                copy.getTags().addAll(newTags);
                copy.getTags().addAll(existingTags);
                copy.getTags().forEach(t -> t.setCreatorWithDate(user));
                copy.getTags().forEach(t -> t.setModifierWithDate(user));
                copy.getQuestionOwners().clear();
                copy.getQuestionOwners().add(user);
                copy.update();
                DB.saveAll(copy.getOptions());
                return new QuestionEntry(question.getId(), copy.getId());
            })
            .toList();
        ArrayNode an = Json.newArray();
        entries.forEach(entry -> an.add(Json.newObject().put("src", entry.srcId).put("dst", entry.dstId)));
        return Results.created((JsonNode) Json.newObject().set("ids", an));
    }
}

File: app/controllers/iop/transfer/impl/DataTransferController.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package controllers.iop.transfer.impl

import controllers.base.scala.ExamBaseController
import io.ebean.DB
import io.ebean.text.PathProperties
import miscellaneous.config.ConfigReader
import miscellaneous.file.FileHandler
import miscellaneous.json.JsonDeserializer
import miscellaneous.scala.{DbApiHelper, JavaApiHelper}
import models.questions.{Question, Tag}
import models.user.{Role, User}
import org.apache.pekko.stream.{IOResult, Materializer}
import org.apache.pekko.stream.scaladsl.{FileIO, Source as ScalaSource}
import org.apache.pekko.util.ByteString
import play.api.Logging
import play.api.libs.Files.TemporaryFile
import play.api.libs.json.*
import play.api.libs.ws.{BodyWritable, InMemoryBody, WSClient}
import play.api.mvc.*
import security.scala.Auth
import security.scala.Auth.{AuthenticatedAction, authorized}
import system.AuditedAction

import java.io.File
import java.net.URI
import java.nio.file.Paths
import javax.inject.Inject
import scala.concurrent.{ExecutionContext, Future}
import scala.jdk.CollectionConverters.*
import scala.util.{Failure, Success, Try}

class DataTransferController @Inject() (
    val controllerComponents: ControllerComponents,
    val authenticated: AuthenticatedAction,
    val audited: AuditedAction,
    wsClient: WSClient,
    configReader: ConfigReader,
    fileHandler: FileHandler
)(implicit ec: ExecutionContext, mat: Materializer)
    extends BaseController
    with DbApiHelper
    with JavaApiHelper
    with Logging:

  private val SeventyMB = 70000 * 1024

  enum DataType:
    case QUESTION

  private case class QuestionEntry(srcId: Long, dstId: Long)

  def importQuestionAttachment(id: Long): Action[MultipartFormData[TemporaryFile]] =
    Action(parse.multipartFormData).andThen(audited) { request =>
      Option(DB.find(classOf[Question], id)) match
        case None => NotFound
        case Some(question) =>
          request.body.file("file") match
            case None => BadRequest("file not found")
            case Some(filePart) =>
              if filePart.fileSize > configReader.getMaxFileSize then BadRequest("i18n_file_too_large")
              else
                Try(copyFile(filePart.ref, "question", id.toString)) match
                  case Failure(_) => InternalServerError("i18n_error_creating_attachment")
                  case Success(newFilePath) =>
                    fileHandler.removePrevious(question)
                    val attachment = fileHandler.createNew(
                      filePart.filename,
                      filePart.contentType.getOrElse("application/octet-stream"),
                      newFilePath
                    )
                    question.setAttachment(attachment)
                    question.save()
                    Created
    }

  def importData(): Action[JsValue] = Action(parse.json(maxLength = SeventyMB)).andThen(audited) { request =>
    (request.body \ "type").asOpt[String] match
      case Some("QUESTION") => importQuestions(request.body)
      case _                => NotAcceptable
  }

  def exportData(): Action[JsValue] =
    authenticated
      .andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN)))
      .andThen(audited)
      .async(parse.json) { request =>
        // Create implicit body writable for JsValue
        given BodyWritable[JsValue] = BodyWritable(
          jsValue => InMemoryBody(ByteString(Json.stringify(jsValue))),
          "application/json"
        )

        val user = request.attrs(Auth.ATTR_USER)
        val body = request.body

        (body \ "type").asOpt[String] match
          case Some("QUESTION") if (body \ "ids").asOpt[JsArray].exists(_.value.nonEmpty) =>
            val ids    = (body \ "ids").as[Seq[Long]].toSet
            val orgRef = (body \ "orgRef").as[String]

            val pp    = PathProperties.parse("(*, options(*), tags(name))")
            val query = DB.find(classOf[Question])
            query.apply(pp)

            val questions = query
              .where()
              .idIn(ids.map(Long.box).asJava)
              .or()
              .eq("questionOwners", user)
              .eq("creator", user)
              .endOr()
              .findSet()
              .asScala
              .toSet

            val data = Json.obj(
              "type"      -> "QUESTION",
              "path"      -> "/integration/iop/import",
              "owner"     -> user.getEppn,
              "questions" -> JsArray(questions.map(q => serialize(q, pp)).toSeq),
              "ids"       -> (body \ "ids").get,
              "orgRef"    -> orgRef
            )

            for
              url       <- parseURL(orgRef)
              uploadUrl <- parseUploadURL(orgRef)
              response  <- wsClient.url(url.toString).post(data)
              result <-
                if response.status != CREATED then
                  Future.successful(
                    InternalServerError((response.json \ "message").asOpt[String].getOrElse("Connection refused"))
                  )
                else
                  val entries = (response.json \ "ids")
                    .as[Seq[JsValue]]
                    .map(id => (id \ "src").as[Long] -> (id \ "dst").as[Long])
                    .toMap

                  val localAttachments = questions
                    .filter(q => Option(q.getAttachment).exists(a => new File(a.getFilePath).exists()))
                    .map(q => q.getId.longValue() -> q.getAttachment)
                    .toMap

                  val remoteAttachments = localAttachments
                    .collect {
                      case (srcId, attachment) if entries.contains(srcId) =>
                        entries(srcId) -> attachment
                    }

                  val uploadFutures = remoteAttachments.map { case (dstId, attachment) =>
                    val host = uploadUrl.replace("/id/", s"/$dstId/")
                    wsClient
                      .url(host)
                      .post(createSource(attachment))
                      .recover { case ex =>
                        logger.error(s"Failed uploading attachment id $dstId", ex)
                        null
                      }
                  }

                  Future.sequence(uploadFutures).map(_ => Created)
            yield result

          case _ => Future.successful(BadRequest)
      }

  private def copyFile(srcFile: TemporaryFile, pathParams: String*): String =
    val newFilePath = fileHandler.createFilePath(pathParams*)
    fileHandler.copyFile(srcFile, new File(newFilePath))
    newFilePath

  private def parseURL(orgRef: String): Future[java.net.URL] =
    Future.fromTry(Try {
      val url = s"${configReader.getIopHost}/api/organisations/$orgRef/export"
      URI.create(url).toURL
    })

  private def parseUploadURL(orgRef: String): Future[String] =
    Future.fromTry(Try {
      val url = s"${configReader.getIopHost}/api/organisations/$orgRef/export/id/attachment"
      URI.create(url).toURL.toString
    })

  private def createSource(
      attachment: models.attachment.Attachment
  ): ScalaSource[MultipartFormData.Part[ScalaSource[ByteString, ?]], ?] =
    val source: ScalaSource[ByteString, Future[IOResult]] = FileIO.fromPath(Paths.get(attachment.getFilePath))
    val filePart = MultipartFormData.FilePart(
      "file",
      attachment.getFileName,
      Some(attachment.getMimeType),
      source
    )
    ScalaSource.single(filePart)

  private def isNewTag(tag: Tag, existing: Seq[Tag]): Boolean =
    !existing.exists(_.getName == tag.getName)

  private def importQuestions(body: JsValue): Result =
    val eppn = (body \ "owner").as[String]
    DB.find(classOf[User]).where().eq("eppn", eppn).find match
      case None => BadRequest("User not recognized")
      case Some(user) =>
        val questions = (body \ "questions").as[Seq[JsValue]]
        val entries = questions.map { questionNode =>
          val question =
            JsonDeserializer.deserialize(classOf[Question], play.libs.Json.parse(Json.stringify(questionNode)))
          val copy = question.copy()
          copy.setParent(null)
          copy.setCreatorWithDate(user)
          copy.setModifierWithDate(user)
          copy.save()

          val userTags     = DB.find(classOf[Tag]).where().eq("creator", user).findList().asScala.toList
          val questionTags = question.getTags.asScala.toList

          val newTags = questionTags.filter(t => isNewTag(t, userTags))
          newTags.foreach(_.setId(null))

          val existingTags = userTags.filter(t => !isNewTag(t, questionTags))

          DB.saveAll(newTags.asJava)
          copy.getTags.addAll(newTags.asJava)
          copy.getTags.addAll(existingTags.asJava)
          copy.getTags.forEach(t => t.setCreatorWithDate(user))
          copy.getTags.forEach(t => t.setModifierWithDate(user))
          copy.getQuestionOwners.clear()
          copy.getQuestionOwners.add(user)
          copy.update()
          DB.saveAll(copy.getOptions)

          QuestionEntry(question.getId, copy.getId)
        }

        val idsArray = JsArray(entries.map(e => Json.obj("src" -> e.srcId, "dst" -> e.dstId)))
        Created(Json.obj("ids" -> idsArray))

  private def serialize(question: Question, pp: PathProperties): JsValue =
    Json.parse(DB.json().toJson(question, pp))
